File: python/146.LRU Cache.py
class Node():
    def __init__(self, key: int, val: int):
        self.key = key
        self.val = val
        self.next = None
        self.pre = None


class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.map = dict()  #key to Node
        self.headNode = Node(-1, -1)
        self.tailNode = Node(-1, -1)
        self.headNode.next = self.tailNode
        self.tailNode.pre = self.headNode
        
    def get(self, key: int) -> int:
        if key not in self.map:
            return -1
        visitNode = self.map[key]
        self.move2tail(visitNode)
        return visitNode.val


    def put(self, key: int, value: int) -> None:
        if key in self.map:
            putNode = self.map[key]
            putNode.val = value
            self.move2tail(putNode)
        else:
            if len(self.map) == self.capacity:
                self.remove()
            insertNode = Node(key, value)
            insertNode.next = self.tailNode
            insertNode.pre = self.tailNode.pre
            self.tailNode.pre.next = insertNode
            self.tailNode.pre = insertNode
            self.map[key] = insertNode
            assert len(self.map) <= self.capacity

    def remove(self):
        removeNode = self.headNode.next
        self.headNode.next = removeNode.next
        removeNode.next.pre = self.headNode
        removeNode.next = None
        removeNode.pre = None
        key = removeNode.key
        del self.map[key]
        del removeNode

    def move2tail(self, node):
        node.pre.next = node.next
        node.next.pre = node.pre

        node.next = self.tailNode
        node.pre = self.tailNode.pre
        self.tailNode.pre.next = node
        self.tailNode.pre = node


if __name__ == '__main__':
    c = LRUCache(1)
    c.put(3, 4)
    c.put(2, 1)
    c.put(5, 6)

File: cpp/146.LRU Cache.cpp
#include <iostream>
#include <stdio.h>
#include <map>
using namespace std;

class Node{
public:
    Node *pre;
    Node *next;
    int key;
    int val;
    Node(int _key, int _val): key(_key), val(_val), pre(NULL), next(NULL){}
//    ~Node(){
//        pre = NULL;
//        next = NULL;
//    }
};


class LRUCache {
private:
    Node *headNode;
    Node *tailNode;
    int cap;
    map<int, Node*>lru_map;
    map<int, Node*>::iterator it;

    void remove2tail(Node *node) {
        node -> pre -> next =  node -> next;
        node -> next -> pre = node -> pre;
        node -> next = tailNode;
        node -> pre = tailNode -> pre;
        tailNode -> pre -> next = node;
        tailNode -> pre = node;
    }
    void removeNode(){
        Node *removeNode = headNode -> next;
        removeNode -> next -> pre = headNode;
        headNode -> next = removeNode -> next;
        lru_map.erase(removeNode->key);
        delete(removeNode);
    }


public:
    LRUCache(int capacity) {
        headNode = new Node(-1, -1);
        tailNode = new Node(-1, -1);
        headNode -> next = tailNode;
        tailNode -> pre = headNode;
        cap = capacity;
    }

    int get(int key) {
        it = lru_map.find(key);
        if (it == lru_map.end())
            return -1;
        else {
            Node *tmpNode = it -> second;
            remove2tail(tmpNode);
            return tmpNode -> val;
        }
    }

    void put(int key, int value) {
        it = lru_map.find(key);
        if (it == lru_map.end()){
            if (lru_map.size() == cap)
                removeNode();
            Node *insertNode = new Node(key, value);
            insertNode -> next = tailNode;
            insertNode -> pre = tailNode -> pre;
            tailNode -> pre -> next = insertNode;
            tailNode -> pre = insertNode;
            lru_map[key] = insertNode;
        }
        else{
            Node *tmpNode = it -> second;
            tmpNode -> val = value;
            remove2tail(tmpNode);
        }
    }
};
